packages/eslint-plugin/src/rules/strict-void-return.ts

```diff
@@ -0,0 +1,428 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+type Options = [
+  {
+    allowReturnAny?: boolean;
+  },
+];
+
+type MessageId = `asyncFunc` | `nonVoidFunc` | `nonVoidReturn`;
+
+export default util.createRule<Options, MessageId>({
+  name: 'strict-void-return',
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Disallow passing a value-returning function in a position accepting a void function',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      asyncFunc:
+        'Async function used in a context where a void function is expected.',
+      nonVoidFunc:
+        'Value-returning function used in a context where a void function is expected.',
+      nonVoidReturn:
+        'Value returned in a context where a void return is expected.',
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false,
+        properties: {
+          allowReturnAny: {
+            type: 'boolean',
+            description:
+              'Whether to allow functions returning `any` to be used in place expecting a `void` function.',
+          },
+        },
+      },
```

Comment:
[Feature] This is what [`@typescript-eslint/no-unsafe-return`](https://typescript-eslint.io/rules/no-unsafe-return) already does, right? I don't think we need this rule to also check for `any` safety - it's not the same area as strict `void` semantics. I think this rule should effectively hardcode `allowReturnAny` to enabled and not let users change that.

```suggestion
```